File: scripts/update-charts.py
#!/usr/bin/env python3

# Fallback port of update-charts.mjs for runners without Node (Spacelift's Alpine
# image ships python3 but no node). Output must stay byte-identical to the .mjs
# version: the chart directory hash in prepare.sh feeds the Helm chart version, so
# any formatting drift shows up as a spurious release upgrade.
# Keep the ignore list and category mapping in sync with update-charts.mjs.

import json
import os
import sys

SCRIPT_NAME = os.path.basename(sys.argv[0])
USAGE = (
    f"Usage: python3 {SCRIPT_NAME} <input-json> [workspace-directory]\n"
    "  <input-json>          Path to a JSON file containing input parameters.\n"
    "  [workspace-directory] Optional path to the repository root. Defaults to current directory."
)

IGNORED_SERVICES = [
    "embassy",
    "prometheus-ecs-discovery",
    "redis-streams-exporter",
    "alb-log-parser",
    "minio",
]

CHART_CATEGORIES = {
    "monitoring": "paragon-monitoring",
    "logging": "paragon-logging",
    "onPrem": "paragon-onprem",
}


def exit_with_error(message):
    print(message, file=sys.stderr)
    sys.exit(1)


def exit_with_usage_error(message):
    exit_with_error(f"{message}\n\n{USAGE}")


def verify_script_args():
    args = sys.argv[1:]
    input_file_path = args[0] if args else None
    workspace_directory = args[1] if len(args) > 1 else "."

    if not input_file_path:
        exit_with_usage_error("Missing required argument: <input.json>")
    if not os.path.exists(input_file_path):
        exit_with_usage_error(f"Specified input file does not exist: {input_file_path}")
    if not os.path.exists(workspace_directory):
        exit_with_usage_error(f"Workspace directory does not exist: {workspace_directory}")
    if not os.path.isdir(os.path.join(workspace_directory, "charts")):
        exit_with_usage_error(
            f"The 'charts' directory was not found in {workspace_directory}. "
            "Please pass the correct path to the repository root as the second argument."
        )

    return input_file_path, workspace_directory


def get_chart_subdir(service_name, service_category):
    if service_name in ("openobserve", "fluent-bit"):
        return CHART_CATEGORIES["logging"]
    if service_category == "monitor":
        return CHART_CATEGORIES["monitoring"]
    return CHART_CATEGORIES["onPrem"]


def write_chart_fixtures(service):
    service_name = service["name"]
    subdir = get_chart_subdir(service_name, service.get("category"))

    flattened = dict(service)
    flattened["envKeys"] = sorted(service["envKeys"])
    flattened["secretKeys"] = sorted(service["secretKeys"])

    print(f"Writing chart fixtures for service: {service_name}")
    chart_directory = os.path.join("charts", subdir, "charts", service_name)
    chart_yaml_path = os.path.join(chart_directory, "Chart.yaml")

    if not os.path.exists(chart_yaml_path):
        print(f"  Helm chart does not exist for service {service_name}!", file=sys.stderr)
        print(f"  Expected path: {chart_yaml_path}", file=sys.stderr)
        print("  Are you missing a helm chart for a newly-added service?", file=sys.stderr)
        print(
            f"  Please create the chart or add the service to the ignore list in {SCRIPT_NAME}.",
            file=sys.stderr,
        )
        return False

    print(f"  Found chart directory: {chart_directory}")
    files_dir = os.path.join(chart_directory, "files")
    os.makedirs(files_dir, exist_ok=True)
    # JSON.stringify(value, null, 2) equivalent: 2-space indent, ": " / ", "
    # separators, literal UTF-8, no trailing newline.
    with open(os.path.join(files_dir, "service-inputs.json"), "w", encoding="utf-8") as handle:
        handle.write(json.dumps(flattened, indent=2, ensure_ascii=False))
    return True


def main():
    input_file_path, workspace_directory = verify_script_args()

    print(f"Generating Helm charts using parameters from: {input_file_path}")
    print(f"Using workspace directory: {os.path.abspath(workspace_directory)}")

    try:
        with open(input_file_path, encoding="utf-8") as handle:
            update_params = json.load(handle)
    except (OSError, ValueError) as error:
        exit_with_error(f"Failed to read or parse input file: {error}. Path: {input_file_path}")

    services = update_params["inputs"]["services"]
    print(
        "Parsed input parameters:",
        {
            "version": update_params.get("version"),
            "platformEnv": update_params.get("platformEnv"),
            "serviceCount": len(services),
        },
    )

    failures = 0
    for service in services:
        if service["name"] in IGNORED_SERVICES:
            print(f"Skipping ignored service: {service['name']}")
            continue
        if not write_chart_fixtures(service):
            failures += 1

    if failures:
        exit_with_error(f"Failed to write chart fixtures for {failures} service(s).")

    print("Successfully wrote chart fixtures for all services.")


main()

File: scripts/update-charts.mjs
#!/usr/bin/env node

// This script accepts a JSON file containing parameters and Paragon
// service information required to update the helm charts.

import { execSync } from "child_process";
import { existsSync, writeFileSync, readFileSync, mkdirSync } from "fs";
import { basename, resolve } from "path";

const SCRIPT_NAME = basename(process.argv[1]);
const USAGE = `Usage: node ${SCRIPT_NAME} <input-json> [workspace-directory]\n` +
  "  <input-json>          Path to a JSON file containing input parameters.\n" +
  "  [workspace-directory] Optional path to the repository root. Defaults to current directory.";

// Add the names of any services to ignore here.
// For any service NOT in this list, a corresponding helm chart
// must exist in the repo.
const ignoredServices = [
  'embassy',
  'prometheus-ecs-discovery',
  'redis-streams-exporter',
  'alb-log-parser',
  // MinIO was retired (PARA-21646); its chart is removed, but atlas input may
  // still list it while datastores/minio remains in the monorepo.
  'minio',
];

/**
 * @typedef {Object} UpdateActionParams
 * @property {string} version
 * @property {string} platformEnv
 * @property {ChartInputs} inputs
 */

/**
 * @typedef {Object} ChartInputs
 * @property {ServiceInfo[]} services
 */

/**
 * @typedef {Object} ServiceInfo
 * @property {string} name
 * @property {string} category
 * @property {string[]} envKeys
 * @property {string[]} secretKeys
 */

const chartCategories = {
  monitoring: { subdir: 'paragon-monitoring' },
  logging: { subdir: 'paragon-logging' },
  onPrem: { subdir: 'paragon-onprem' },
}

async function main() {
  const { inputFilePath, workspaceDirectory } = verifyScriptArgs();
  process.cwd(workspaceDirectory);

  console.log(`Generating Helm charts using parameters from: ${inputFilePath}`);
  console.log(`Using workspace directory: ${resolve(workspaceDirectory)}`);
  console.log(`Code branch: ${getCurrentGitBranch()}`);

  const updateParams = readInputJSONFile(inputFilePath);
  const { version, platformEnv, inputs } = updateParams;
  console.log('Parsed input parameters:', { version, platformEnv, serviceCount: inputs.services.length });

  let failures = 0;
  inputs.services.forEach(service => {
    if (ignoredServices.includes(service.name)) {
      console.log(`Skipping ignored service: ${service.name}`);
      return;
    }
    if (!writeChartFixtures(service)) {
      failures += 1;
    }
  });

  if (failures > 0) {
    exitWithError(`Failed to write chart fixtures for ${failures} service(s).`);
  }

  console.log('Successfully wrote chart fixtures for all services.');
}

/**
 * Read and verify CLI arguments passed to the script.
 * @returns { inputFilePath: string, workspaceDirectory: string }
 */
function verifyScriptArgs() {
  const [inputFilePath, workspaceDirectory = '.'] = process.argv.slice(2);

  if (!inputFilePath) {
    exitWithUsageError('Missing required argument: <input.json>');
  }

  if (!existsSync(inputFilePath)) {
    exitWithUsageError(`Specified input file does not exist: ${inputFilePath}`);
  }

  if (!existsSync(workspaceDirectory)) {
    exitWithUsageError(`Workspace directory does not exist: ${workspaceDirectory}`);
  }

  if (!existsSync(`${workspaceDirectory}/charts/.`)) {
    exitWithUsageError(`The 'charts' directory was not found in ${workspaceDirectory}. Please pass the correct path to the repository root as the second argument.`);
  }

  return { inputFilePath, workspaceDirectory }
}

/**
 * @return {never}
 */
function exitWithError(message) {
  console.error(message);
  process.exit(1);
}

function exitWithUsageError(message) {
  exitWithError(message + "\n\n" + USAGE);
}

/**
 * @param {string} filePath 
 * @returns {UpdateActionParams}
 */
function readInputJSONFile(filePath) {
  try {
    return JSON.parse(readFileSync(filePath, 'utf-8'));
  } catch (error) {
    exitWithError(`Failed to read or parse input file: ${error.message}. Path: ${filePath}`);
  }
}

/**
 * @param {ServiceInfo} service
 * @return {boolean}
 */
function writeChartFixtures(service) {
  const { name: serviceName, category: serviceCategory, envKeys, secretKeys } = service;
  const chartCategory = getChartCategory(serviceName, serviceCategory);

  const flattenedService = {
    ...service,
    envKeys: Object.keys(envKeys).sort(),
    secretKeys: Object.keys(secretKeys).sort(),
  }

  // TODO: This is just a stub. Implement actual chart fixture writing logic here.
  console.log(`Writing chart fixtures for service: ${serviceName}`);
  const chartDirectory = `charts/${chartCategory.subdir}/charts/${serviceName}`;
  const chartYamlPath = `${chartDirectory}/Chart.yaml`;
  // e.g. charts/paragon-onprem/charts/zeus/Chart.yaml

  if (existsSync(chartYamlPath)) {
    console.log(`  Found chart directory: ${chartDirectory}`);
    mkdirSync(`${chartDirectory}/files`, { recursive: true });
    writeFileSync(`${chartDirectory}/files/service-inputs.json`, JSON.stringify(flattenedService, null, 2));
    return true;
  } else {
    console.error(`  Helm chart does not exist for service ${serviceName}!`);
    console.error(`  Expected path: ${chartYamlPath}`);
    console.error(`  Are you missing a helm chart for a newly-added service?`);
    console.error(`  Please create the chart or add the service to the ignore list in ${SCRIPT_NAME}.`);
    return false;
  }
}

function getChartCategory(serviceName, serviceCategory) {
  if (['openobserve', 'fluent-bit'].includes(serviceName)) {
    return chartCategories.logging;
  }
  if (serviceCategory === 'monitor') {
    return chartCategories.monitoring;
  }
  return chartCategories.onPrem;
}

function getCurrentGitBranch() {
  try {
    return execSync('git branch --show-current 2>/dev/null', { encoding: 'utf-8' }) || '(none)';
  } catch {
    return '(unavailable)';
  }
}

main().catch(error => {
  exitWithError(`Unexpected error: ${error.message} `);
});